### PR TITLE
Assign permissions - backend

### PR DESCRIPTION
**Pull request by jasalisbury**
**Description of change**

I'll get another PR up with frontend changes that merges into this branch to make both easier to review.

 * Add user seeder
 * add test users as a seeder
 * created at and updated at for users and permissions have default values of "now"
 * users "title" column renamed to "role"
 * GET `/user` returns the currently logged in user from the DB

**Notes**

 * Would be nice to have a validation on permissions that asserts any "global" scope id (the admin scope, "2") has to have a region id of 14 (no region). I tried to get a [model validation](https://sequelize.org/v5/manual/models-definition.html#model-wide-validations) working but something wasn't working correctly. Either need a ticket to write the validation or need to pair with someone to figure it out.

**Questions**
 * How do we run migrations/seeders in sandbox and dev?

**How to test**

Run tests/linter/cucumber tests

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/52

**Checklist**

- [-] Meets issue criteria
- [x] Code tested
- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

File: src/migrations/20201130184519-rename-user-title-column.js
Comment (kryswisnaskas) [inline src/migrations/20201130184519-rename-user-title-column.js:50]: Renaming to "role" makes sense. This might also change going forward, since we are introducing the "Role" table. If sticking with enum for now, we will also need to drop the enum type. Otherwise, there will be an error trying to undo the migration.

Something similar to this can be used (although untested in migrations):
```
const query = 'DELETE FROM pg_type WHERE typname = \'enum_Users_title\' ';
    return queryInterface.sequelize.query(query);
```
The same goes for `enum_Users_role`

File: src/migrations/20201124173716-add-default-createdat-updatedat.js
Comment (kryswisnaskas) [inline src/migrations/20201124173716-add-default-createdat-updatedat.js:5]: We might not need this. Sequelize already handles the timestamps, unless explicitly defined in the model "timestamps: false".
The only reason I can think of this would be useful is if we needed to bypass `sequelize` for some reason. But even if that was the case we could manually add the dates. Is there anything I am missing?

File: src/migrations/20201130184519-rename-user-title-column.js
Comment (kryswisnaskas) [inline src/migrations/20201130184519-rename-user-title-column.js:31]: Can we add a transaction if there are multiple statements in one migration? The main reason is that in a case of one statement failing and other succeeding, potentially leaves us in an undesired state.

File: src/seeders/20201124160449-users.js
Comment (kryswisnaskas) [inline src/seeders/20201124160449-users.js:111]: Is the user import via seeding designed for non-prod environments only?
Even though `sequelize `offers seeders, we might want to substitute them with migrations for production. The main reason is that with seeders we can't run them multiple times.

File: src/routes/apiDirectory.js
Comment (kryswisnaskas) [inline src/routes/apiDirectory.js:25]: Can we wrap these 2 lines (25-26) into a `try catch` - mainly to prevent calling `toJSON()` if the `user `is `null `for some reason

**Review by kryswisnaskas — COMMENTED** [review]
It works! 🎉 I was able to add permissions to a user and see them both in the db and in the UI.

Left a few comments mainly about the migrations.

File: src/migrations/20201124173716-add-default-createdat-updatedat.js
Comment (jasalisbury) [reply src/migrations/20201124173716-add-default-createdat-updatedat.js:5]: Hum whenever I don't change those columns I get `ERROR: null value in column "createdAt" violates not-null constraint` running the seeder. Maybe Sequelize doesn't set the timestamp columns when doing a `bulkInsert`?

File: src/seeders/20201124160449-users.js
Comment (jasalisbury) [reply src/seeders/20201124160449-users.js:111]: Yea the user import should only be used in non-prod environments. My thinking was use seeders for local/non-prod environments (because seeding the DB locally will probably be a common task as we get more test data) and custom scripts for loading up data in prod (because that will be a "one time" occurrence). There may be a better way I haven't thought of, let me know!

File: .circleci/config.yml
Comment (SarahJaine) [inline .circleci/config.yml:220]: Instead of naming these variables twice, I think you can [use the `environment` key at the job level](https://circleci.com/docs/2.0/env-vars/?utm_medium=SEM&utm_source=gnb&utm_campaign=SEM-gb-DSA-Eng-uscan&utm_content=&utm_term=dynamicSearch-&gclid=Cj0KCQiA2af-BRDzARIsAIVQUOfLm5QJc1iQTMOgJFBztJ7z-GQxZNrk0dAxhx-OwUyFDeglv5J3HDcaAsRlEALw_wcB#setting-an-environment-variable-in-a-job) and thereby supply them once.

File: package.json
Comment (SarahJaine) [inline package.json:34]: Thanks for adding both docker and local commands.

File: src/migrations/20201124173716-add-default-createdat-updatedat.js
Comment (SarahJaine) [reply src/migrations/20201124173716-add-default-createdat-updatedat.js:5]: Weird- createdAt and updatedAt are supposed to be managed by sequelize. https://sequelize.org/master/manual/model-basics.html#timestamps  Bulk operations shouldn't have an effect on this. I don't want to hold this PR up, but I would consider this tech debt that we should address later if it gets merged as is.

File: src/routes/admin/user.js
Comment (SarahJaine) [inline src/routes/admin/user.js:50]: Love this change

File: src/seeders/20201124160449-users.js
Comment (SarahJaine) [inline src/seeders/20201124160449-users.js:58]: It seems weird to me to set the pks since they're an auto-incrementing id. It makes it so that you'd have to dump the db and reset ids on the table before running seed. Is this necessary?

**Review by SarahJaine — COMMENTED** [review]
Looks great. 🙌  Just a few questions and suggestions.

File: src/seeders/20201124160449-users.js
Comment (jasalisbury) [reply src/seeders/20201124160449-users.js:58]: yeah we can drop the explicit id for most users but will need to keep an id for at least one user so we have a known id when `SKIP_AUTH` is true.

Comment (jasalisbury) [reply src/seeders/20201124160449-users.js:58]: We could also just use a well known name and remove all ids

Comment (SarahJaine) [reply src/seeders/20201124160449-users.js:58]: Instead of using a well known name, can we look up the user by `hsesUserId` or `email` since those are both unique?

Comment (jasalisbury) [reply src/seeders/20201124160449-users.js:58]: Wouldn't we run into the same problem using a unique field that we would if we used `id`, you'd have to dump the db and reset the unique field before seeding?

After thinking about this a bit we may want to keep using explicit IDs for seeds. In the future if we want to add additional seeders (for activity reports for example) we will need to tie those records to a user. Seems easier to have a few well known user ids we can use to tie everything together. Or we can create users in each new seed file? Might end up with a lot of users but probably doesn't matter that much.

Another option would be to use something like [Factory girl](https://github.com/simonexmachina/factory-girl) to handle all the dynamic associations for us. If we went that route I imagine we'd end up with a single seeder file that handles creating all our testing data. I've used this approach in the past with some success, granted it was with rails (using [Factory bot](https://github.com/thoughtbot/factory_bot)). Another advantage of using a factory library is it would probably speed up testing on the backend.

I'm not really sold on any approach right now, interested to hear other's thoughts.

Comment (kryswisnaskas) [reply src/seeders/20201124160449-users.js:58]: I am fine with either approach. Seems a bit easier/clearer to use explicit ids. The testing speed is a consideration, though. If you think the benefits would be substantial we could go that route.

**Review by kryswisnaskas — APPROVED** [review]
Looks good! 👍

**Comment by jasalisbury** [comment]
Waiting for https://github.com/HHS/Head-Start-TTADP/issues/202 before merging in so we can run the migrations and seeders once merged.

File: .circleci/config.yml
Comment (kryswisnaskas) [inline .circleci/config.yml:115]: Shouldn't this be just 'yarn db:seed'?

Comment (jasalisbury) [reply .circleci/config.yml:115]: yep, one of the reasons the latest push failed to deploy ha...

**Comment by rahearn** [comment]
Is this ready to get merged?

**Pull request merged**
